About Project (#43)
scolladon/lwcc: docs/general/about.md

# About the project

## The team

### [Vic](https://github.com/victorgz)
Spanish IT Engineer & Business  
Salesforce Technical Consultant  
Innovation and technology enthusiast  

### [Seb](https://github.com/scolladon)
French IT Engineer  
Full Stack Salesforce Architect (CTA)  
Technology passionated  

### [PM](https://github.com/pmdartus)

> 

Reviewer: ```suggestion
Principal Software Engineer - Salesforce UI platform
LWC & JSDOM maintainer
```

We have different kind of backgrounds with web technology.
Some of us are better in back and keen on those new web technology.

It was the opportunity to share and build deep knowledge with skills acquired by resolving challenging situations.
We wanted the quality to be the best possible with the few time we can spend on it because of our daily jobs

### Why chart.js

When we started the project to bring charts in LWC we needed to choose a path to reach the goal.

Starting from scratch would have been very challenging, as we don't have any knowledge in that field and as it is a huge task we don't really considered this. It was tacitly acknowledged we would not go this road :D

Then we needed to identify which kind of library we could use, knowing we cannot use a library manipulating another namespace than HTML (Lightning Web Component constraints).

After having reviewed js charting libraries our choice has converge on Chart.js because of the ecosystem and other existing porting we [trust](https://github.com/chartjs/awesome#javascript).
We were able to create a concluding POC in 7 minutes using Chart.js and LWC !

> ### Contribute to the eco system

Reviewer: ```suggestion
### Contribute to the ecosystem
```

We are all Salesforce employees and we want to improve our ecosystem as much as possible.
We are all fervant believers of the LWC technology and we wanted to demonstrate how powerful and convenient it is to use it.
We believe bringing awesome tools and libraries will encourage awesome people to create awesome values to their customers using it.

## Challenges encountered

### Coding convention

As in every project, we knew we would face this challenge at some point.
In order to anticipate this we have setup very early a workshop to define the coding convention.

It ended up with prettier and linter with the right plugins (apex and lwc) and a naming convention for the property our components will expose.
We use husky to ensure the quality level of the code is pretty much the same everytime.
We also have implemented CI to further industrialize those detection.
We put in place Github Pull Request template to allow answering the right question when creating PR.

### Migrate from local project to Salesforce project

We started to build the project using LWC open source in local.
It was very convenient because we had automatic reload and easy debugging. We won lot of time.

> When the project has been mature enough and we had solved most of our big problems we decided to move to LWC Salesforce.

Reviewer: ```suggestion
When the project was mature enough and we had solved most of our big problems we decided moving the local project to a Salesforce plaform project.
```

> First problem we encountered was the necessity to structure all the component in the same sub directory.

Reviewer: ```suggestion
The first problem we encountered was the necessity to restructure all the components in the same subdirectory.
```

> It is not possible to have a logical structure to group all your component logically as you would do in any project.

Reviewer: ```suggestion
The platform format makes it impossible to have a logical folder structure to group all your components as you would do in any project.
```

> In Salesforce every component folder are stored in the lwc sub folder of the sources.

Reviewer: ```suggestion
On Salesforce every component folder is stored in the `lwc` subfolder of the sources.
```

> While migrating in LWC Salesforce we discovered template requires to have an explicit slot to be able to have children component communicating with each other (Locker Service) while composing them, which is not the case in LWC OS. We added slot on every component.

Reviewer: ```suggestion
While migrating in LWC Salesforce we discovered that templates require to have an explicit slot to be able to have children component communicating with each other (Locker Service) while composing them, which is not the case in LWC OS. We added the `<slot>` in every component.
```

We use the connected callback in order to attach listeners to the parent element allowing children to "bubble" up information they gather via property using dispatchEvent. When a component property is updated it just dispatches the change and the parent catches it. When a component disconnects (`disconnectedCallback`) it fetches its parent's reference stored at the `connectedCallback` and dispatches an event to the parent so it can spread the deletion and update the chart.

#### Property reactivity

In order to detect property changes and to dispatch an event to the parent we used getter and setter properties. In the setter of the property we dispatch an event to the parent to let it know a property has been updated and decide what to do.
To simplify the work and avoid to duplicate the code we created a reactivity manager component which serves a Proxy object catching the setter to call a callback method in this case. The Proxy object was used in each property to store value exposed via getter setter. This way we had one javascript object in our component to handle the reactivity and expose/store properties.

> #### Reactiviy throttling

Reviewer: ```suggestion
#### Reactivity throttling
```
